torch/csrc/jit/codegen/cuda/expr_simplifier.cpp

```diff
@@ -833,14 +941,83 @@ Val* eliminateTrivialComputation(Val* value) {
   return value;
 }
 
+// If x can be proved to be non-negative, then replace x >= 0 as true, replace
+// x < 0 as false
+// If x can be proved to be positive, then replace x >= 0 and x > 0 as true,
+// replace x <= 0 and x < 0 as false
+// If x can be proved to be nonzero, then replace x != 0 as true, replace x == 0
+// as false
+// if x->sameAs(y), then replace x == y as true, replace x != y as false
+Val* eliminateTrivialPredicate(Val* value, const VarInfo& var_info) {
+  if (value->isABool()) {
+    if (auto bop = dynamic_cast<BinaryOp*>(value->definition())) {
+      auto op = bop->getBinaryOpType();
+      if (bop->lhs()->sameAs(bop->rhs())) {
+        if (op == BinaryOpType::Eq) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (op == BinaryOpType::NE) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        }
+      }
+      if (bop->rhs()->isZeroInt()) {
+        if (op == BinaryOpType::GE &&
+            prove::isNonNegative(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::GT && prove::isPositive(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::NE && prove::isNonZero(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(true, DataType::Bool);
+        } else if (
+            op == BinaryOpType::LT &&
+            prove::isNonNegative(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        } else if (
+            op == BinaryOpType::LE && prove::isPositive(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        } else if (
+            op == BinaryOpType::Eq && prove::isNonZero(bop->lhs(), var_info)) {
+          return IrBuilder::newConstant(false, DataType::Bool);
+        }
+      } else if (bop->lhs()->isZeroInt()) {
```

Comment:
nit: Looks like the block below is mostly a duplicate of the above block. Would be great if refactored.

Reply:
I can not think of a more readable way to refactor this. This is like implementing a table lookup using `if` statements. Different branches will look similar, but each with some small modifications. To me, the most straightforward way to implement this is to just list them all.